hazelcast/src/main/java/com/hazelcast/map/impl/query/QueryRunner.java

```diff
@@ -68,6 +75,38 @@ public QueryRunner(MapServiceContext mapServiceContext, QueryOptimizer optimizer
         this.resultProcessorRegistry = resultProcessorRegistry;
     }
 
+    /**
+     * Runs a query on a chunk of a single partition. The chunk is defined by the offset {@code tableIndex}
+     * and the soft limit {@code fetchSize}.
+     *
+     * @param query       the query
+     * @param partitionId the partition which is queried
+     * @param tableIndex  the index at which to start querying
+     * @param fetchSize   the soft limit for the number of items to be queried
+     * @return the queryied entries along with the next {@code tableIndex} to resume querying
+     */
+    public MapEntriesWithCursor runPartitionScanQueryOnPartitionChunk(Query query, int partitionId, int tableIndex,
+                                                                      int fetchSize) {
+        final MapContainer mapContainer = mapServiceContext.getMapContainer(query.getMapName());
+        final Predicate predicate = queryOptimizer.optimize(query.getPredicate(), mapContainer.getIndexes());
+        final QueryableEntriesWithCursor entriesWithCursor =
+                partitionScanExecutor.executeOnPartitionChunk(query.getMapName(), predicate, partitionId, tableIndex, fetchSize);
+
+        final Collection<QueryableEntry> entries = entriesWithCursor.getEntries();
+
+        updateStatistics(mapContainer);
+
+        final Projection projection = query.getProjection();
+        final ArrayList<Entry<Data, Data>> projected = new ArrayList<Entry<Data, Data>>(entries.size());
+
+        for (QueryableEntry entry : entries) {
+            final Data value = serializationService.toData(projection.transform(entry));
+            projected.add(new SimpleEntry<Data, Data>(entry.getKeyData(), value));
```

Comment:
I tried that but I reverted it to this since I wanted the predicate and projection to run (meaning the query needs to be `IterationType#VALUE`) but it to return entries because the iterator needs both values to return to the user and keys for `AbstractMapPartitionIterator#remove` (meaning the query would then be `IterationType#ENTRY`). 
It's a bit of a mess so I chose this since it didn't involve too much duplication. I would prefer to avoid it so if you have a suggestion please say so.